File: java/src/main/java/io/github/emindeniz99/applepurchase/receipt/VerifyReceiptEndpoint.java
package io.github.emindeniz99.applepurchase.receipt;

import io.github.emindeniz99.applepurchase.VerificationException;
import io.github.emindeniz99.applepurchase.VerificationException.Reason;

import java.security.cert.X509Certificate;
import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Drop-in local replacement for Apple's deprecated {@code verifyReceipt}
 * endpoint: same request body, same response body shape, same status codes —
 * but verified offline against the pinned Apple root instead of by calling
 * Apple. Field-by-field fidelity and the unavoidable gaps (fields that only
 * exist in Apple's server-side subscription database, like
 * {@code latest_receipt_info} / {@code pending_renewal_info}) are documented
 * in COMPARISON.md.
 *
 * <p>Like Apple's endpoint, this does NOT check the bundle id — the caller
 * compares {@code receipt.bundle_id}, exactly as with the real endpoint.</p>
 */
public final class VerifyReceiptEndpoint {

    public static final int STATUS_OK = 0;
    /** Malformed request or receipt-data property. */
    public static final int STATUS_MALFORMED = 21002;
    /** Receipt could not be authenticated. */
    public static final int STATUS_NOT_AUTHENTICATED = 21003;
    /** Sandbox receipt sent to the production environment. */
    public static final int STATUS_SANDBOX_RECEIPT_ON_PRODUCTION = 21007;
    /** Production receipt sent to the sandbox environment. */
    public static final int STATUS_PRODUCTION_RECEIPT_ON_SANDBOX = 21008;
    /** Internal error. */
    public static final int STATUS_INTERNAL = 21009;

    private static final DateTimeFormatter FORMAT =
            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
    private static final ZoneId GMT = ZoneId.of("UTC");
    private static final ZoneId PACIFIC = ZoneId.of("America/Los_Angeles");

    private final ReceiptVerifier verifier;
    private final boolean production;

    /**
     * @param trustedRoots pinned roots (production:
     *                     {@code AppleRootCerts.receiptRoots()})
     * @param production   which environment this instance emulates
     *                     (drives 21007/21008 routing)
     */
    public VerifyReceiptEndpoint(Set<X509Certificate> trustedRoots, boolean production) {
        // The bundle id is never consulted: verifyCore skips the claim check,
        // matching Apple's endpoint (callers compare receipt.bundle_id).
        this.verifier = new ReceiptVerifier(trustedRoots, "*");
        this.production = production;
    }

    /**
     * Handles one verifyReceipt request body. Never throws — like the real
     * endpoint, failures are reported through {@code status}.
     */
    public Map<String, Object> verifyReceipt(Map<String, ?> requestBody) {
        Object receiptData = requestBody == null ? null : requestBody.get("receipt-data");
        if (!(receiptData instanceof String) || ((String) receiptData).isEmpty()) {
            return status(STATUS_MALFORMED);
        }
        byte[] der;
        try {
            der = Base64.getMimeDecoder().decode((String) receiptData);
        } catch (IllegalArgumentException e) {
            return status(STATUS_MALFORMED);
        }
        AppReceipt receipt;
        try {
            receipt = verifier.verifyCore(der);
        } catch (VerificationException e) {
            return status(e.reason() == Reason.INVALID_RECEIPT_FORMAT
                    ? STATUS_MALFORMED : STATUS_NOT_AUTHENTICATED);
        } catch (RuntimeException e) {
            return status(STATUS_INTERNAL);
        }

        // 21007/21008 environment routing from the receipt_type attribute.
        // Production types are exactly "Production" and "ProductionVPP";
        // everything else ("ProductionSandbox", "ProductionVPPSandbox",
        // "Xcode", or a missing attribute) fails closed as non-production.
        boolean productionReceipt = "Production".equals(receipt.receiptType())
                || "ProductionVPP".equals(receipt.receiptType());
        if (production && !productionReceipt) {
            return status(STATUS_SANDBOX_RECEIPT_ON_PRODUCTION);
        }
        if (!production && productionReceipt) {
            return status(STATUS_PRODUCTION_RECEIPT_ON_SANDBOX);
        }

        Map<String, Object> response = new LinkedHashMap<String, Object>();
        response.put("status", STATUS_OK);
        response.put("environment", production ? "Production" : "Sandbox");
        response.put("receipt", receiptJson(receipt, Instant.now()));
        return response;
    }

    private static Map<String, Object> status(int code) {
        Map<String, Object> response = new LinkedHashMap<String, Object>();
        response.put("status", code);
        return response;
    }

    private static Map<String, Object> receiptJson(AppReceipt receipt, Instant requestDate) {
        Map<String, Object> json = new LinkedHashMap<String, Object>();
        put(json, "receipt_type", receipt.receiptType());
        put(json, "bundle_id", receipt.bundleId());
        put(json, "application_version", receipt.appVersion());
        put(json, "original_application_version", receipt.originalAppVersion());
        appleDates(json, "receipt_creation_date", receipt.creationDate());
        appleDates(json, "request_date", requestDate);
        appleDates(json, "original_purchase_date", receipt.originalPurchaseDate());
        appleDates(json, "expiration_date", receipt.expirationDate());
        List<Map<String, Object>> inApp = new ArrayList<Map<String, Object>>();
        for (InAppPurchase purchase : receipt.inAppPurchases()) {
            inApp.add(inAppJson(purchase));
        }
        json.put("in_app", inApp);
        return json;
    }

    private static Map<String, Object> inAppJson(InAppPurchase purchase) {
        Map<String, Object> json = new LinkedHashMap<String, Object>();
        put(json, "quantity", stringOrNull(purchase.quantity()));
        put(json, "product_id", purchase.productId());
        put(json, "transaction_id", purchase.transactionId());
        put(json, "original_transaction_id", purchase.originalTransactionId());
        appleDates(json, "purchase_date", purchase.purchaseDate());
        appleDates(json, "original_purchase_date", purchase.originalPurchaseDate());
        appleDates(json, "expires_date", purchase.expiresDate());
        appleDates(json, "cancellation_date", purchase.cancellationDate());
        put(json, "web_order_line_item_id", stringOrNull(purchase.webOrderLineItemId()));
        if (purchase.isInIntroOfferPeriod() != null) {
            json.put("is_in_intro_offer_period",
                    String.valueOf(purchase.isInIntroOfferPeriod() == 1L));
        }
        return json;
    }

    private static String stringOrNull(Long value) {
        return value == null ? null : String.valueOf(value);
    }

    private static void put(Map<String, Object> json, String key, Object value) {
        if (value != null) {
            json.put(key, value);
        }
    }

    /** Apple's three date renderings: {@code x} (GMT), {@code x_ms}, {@code x_pst}. */
    private static void appleDates(Map<String, Object> json, String prefix, Instant instant) {
        if (instant == null) {
            return;
        }
        json.put(prefix, FORMAT.format(instant.atZone(GMT)) + " Etc/GMT");
        json.put(prefix + "_ms", String.valueOf(instant.toEpochMilli()));
        json.put(prefix + "_pst", FORMAT.format(instant.atZone(PACIFIC)) + " America/Los_Angeles");
    }
}

File: python/apple_purchase_verifier/verify_receipt_endpoint.py
"""Drop-in local replacement for Apple's deprecated ``verifyReceipt``
endpoint: same request body, same response body shape, same status codes —
but verified offline against the pinned Apple root instead of by calling
Apple. Field-by-field fidelity and the unavoidable gaps (fields that only
exist in Apple's server-side subscription database, like
``latest_receipt_info`` / ``pending_renewal_info``) are documented in
COMPARISON.md.

Like Apple's endpoint, this does NOT check the bundle id — the caller
compares ``receipt["bundle_id"]``, exactly as with the real endpoint."""

import base64
import binascii
from datetime import datetime, timezone
from typing import Any, Dict, List, Mapping, Optional
from zoneinfo import ZoneInfo

from .exceptions import Reason, VerificationError
from .receipt import AppReceipt, InAppPurchase, verify_receipt_core

STATUS_OK = 0
#: Malformed request or receipt-data property.
STATUS_MALFORMED = 21002
#: Receipt could not be authenticated.
STATUS_NOT_AUTHENTICATED = 21003
#: Sandbox receipt sent to the production environment.
STATUS_SANDBOX_RECEIPT_ON_PRODUCTION = 21007
#: Production receipt sent to the sandbox environment.
STATUS_PRODUCTION_RECEIPT_ON_SANDBOX = 21008
#: Internal error.
STATUS_INTERNAL = 21009

_PACIFIC = ZoneInfo("America/Los_Angeles")


class VerifyReceiptEndpoint:
    """One instance emulates one environment (drives 21007/21008 routing).

    :param trusted_roots: pinned roots
        (production: :func:`apple_purchase_verifier.apple_receipt_roots`)
    :param environment: ``"Production"`` or ``"Sandbox"``
    """

    def __init__(self, trusted_roots, environment: str):
        roots = list(trusted_roots)
        if not roots:
            raise ValueError("trusted_roots must not be empty")
        if environment not in ("Production", "Sandbox"):
            raise ValueError("environment must be 'Production' or 'Sandbox'")
        self._roots = roots
        self._production = environment == "Production"

    def verify_receipt(self, request_body: Optional[Mapping[str, Any]]) -> Dict[str, Any]:
        """Handles one verifyReceipt request body. Never raises — like the
        real endpoint, failures are reported through ``status``."""
        receipt_data = request_body.get("receipt-data") if isinstance(request_body, Mapping) \
            else None
        if not isinstance(receipt_data, str) or not receipt_data:
            return {"status": STATUS_MALFORMED}
        try:
            der = base64.b64decode(receipt_data)
        except (binascii.Error, ValueError):
            return {"status": STATUS_MALFORMED}
        try:
            fields = verify_receipt_core(der, self._roots)
        except VerificationError as e:
            if e.reason == Reason.INVALID_RECEIPT_FORMAT:
                return {"status": STATUS_MALFORMED}
            return {"status": STATUS_NOT_AUTHENTICATED}
        except Exception:
            return {"status": STATUS_INTERNAL}

        # 21007/21008 environment routing from the receipt_type attribute.
        # Production types are exactly "Production" and "ProductionVPP";
        # everything else ("ProductionSandbox", "ProductionVPPSandbox",
        # "Xcode", or a missing attribute) fails closed as non-production.
        production_receipt = fields.receipt_type in ("Production", "ProductionVPP")
        if self._production and not production_receipt:
            return {"status": STATUS_SANDBOX_RECEIPT_ON_PRODUCTION}
        if not self._production and production_receipt:
            return {"status": STATUS_PRODUCTION_RECEIPT_ON_SANDBOX}
        return {
            "status": STATUS_OK,
            "environment": "Production" if self._production else "Sandbox",
            "receipt": _receipt_json(fields, datetime.now(timezone.utc)),
        }


def _receipt_json(fields: AppReceipt, request_date: datetime) -> Dict[str, Any]:
    receipt: Dict[str, Any] = {}
    _put(receipt, "receipt_type", fields.receipt_type)
    _put(receipt, "bundle_id", fields.bundle_id)
    _put(receipt, "application_version", fields.app_version)
    _put(receipt, "original_application_version", fields.original_app_version)
    _apple_dates(receipt, "receipt_creation_date", fields.creation_date)
    _apple_dates(receipt, "request_date", request_date)
    _apple_dates(receipt, "original_purchase_date", fields.original_purchase_date)
    _apple_dates(receipt, "expiration_date", fields.expiration_date)
    receipt["in_app"] = [_in_app_json(p) for p in fields.in_app_purchases]
    return receipt


def _in_app_json(purchase: InAppPurchase) -> Dict[str, Any]:
    entry: Dict[str, Any] = {}
    _put(entry, "quantity", _str_or_none(purchase.quantity))
    _put(entry, "product_id", purchase.product_id)
    _put(entry, "transaction_id", purchase.transaction_id)
    _put(entry, "original_transaction_id", purchase.original_transaction_id)
    _apple_dates(entry, "purchase_date", purchase.purchase_date)
    _apple_dates(entry, "original_purchase_date", purchase.original_purchase_date)
    _apple_dates(entry, "expires_date", purchase.expires_date)
    _apple_dates(entry, "cancellation_date", purchase.cancellation_date)
    _put(entry, "web_order_line_item_id", _str_or_none(purchase.web_order_line_item_id))
    if purchase.is_in_intro_offer_period is not None:
        entry["is_in_intro_offer_period"] = \
            "true" if purchase.is_in_intro_offer_period == 1 else "false"
    return entry


def _str_or_none(value: Optional[int]) -> Optional[str]:
    return None if value is None else str(value)


def _put(target: Dict[str, Any], key: str, value: Any) -> None:
    if value is not None:
        target[key] = value


def _apple_dates(target: Dict[str, Any], prefix: str, date: Optional[datetime]) -> None:
    """Apple's three date renderings: ``x`` (GMT), ``x_ms``, ``x_pst``."""
    if date is None:
        return
    utc = date.astimezone(timezone.utc)
    target[prefix] = utc.strftime("%Y-%m-%d %H:%M:%S") + " Etc/GMT"
    target[prefix + "_ms"] = str(int(utc.timestamp() * 1000))
    target[prefix + "_pst"] = \
        utc.astimezone(_PACIFIC).strftime("%Y-%m-%d %H:%M:%S") + " America/Los_Angeles"
